### ISSUE TITLE
`no-test-callback` doesn't handle `done` being used to fail a test

### ISSUE THREAD

**Issue by SimenB** [jest-community/eslint-plugin-jest#223]
The `done` callback in Jest can be used to fail tests, either by `done.fail()` or passing anything into `done`, like this: `done('fail')`. Both of these cases are handled wrong by the rule now - always replacing `done` with `resolve, which only works when it's not supposed to fail.

I've added 2 failing tests:

```diff
diff --git i/rules/__tests__/no-test-callback.test.js w/rules/__tests__/no-test-callback.test.js
index 277359d..2b52cf6 100644
--- i/rules/__tests__/no-test-callback.test.js
+++ w/rules/__tests__/no-test-callback.test.js
@@ -123,5 +123,29 @@ ruleTester.run('no-test-callback', rule, {
       output:
         'test("something", async function () {await new Promise((done) => {done();})})',
     },
+    {
+      code: 'test("something", done => {done("fail");})',
+      errors: [
+        {
+          message: 'Illegal usage of test callback',
+          line: 1,
+          column: 19,
+        },
+      ],
+      output:
+        'test("something", () => {return new Promise((_, reject) => {reject("error");})})',
+    },
+    {
+      code: 'test("something", done => {done.fail("fail");})',
+      errors: [
+        {
+          message: 'Illegal usage of test callback',
+          line: 1,
+          column: 19,
+        },
+      ],
+      output:
+        'test("something", () => {return new Promise((_, reject) => {reject("error");})})',
+    },
   ],
 });
```

**Comment by SimenB** [comment]
Bad example, should be

```js
test('some test', done => {
  setTimeout(() => {
    expect(false).toBe(true);
    done();
  }, 10);
});
```

**Comment by aryehb** [comment]
I just tested that as well with the same results. The error is immediately caught by Jest without waiting for the timeout.

**Comment by astorije** [comment]
@SimenB's snippet above makes sense, however @aryehb's latest comment is accurate: `10` is in milliseconds and therefore shorter than the default 5s timeout. But Jest does handle the test failure correctly even in the situation above.

The following snippet does timeout like the documentation of `no-test-callback` warns about:

```ts
test('some test', done => {
  setTimeout(() => {
    expect(false).toBe(true);
    done();
  }, 10000);
});
```

With the following output:

```ts
  ● some test

    : Timeout - Async callback was not invoked within the 5000ms timeout specified by jest.setTimeout.Timeout - Async callback was not invoked within the 5000ms timeout specified by jest.setTimeout.Error:

    > 10 | test('some test', done => {
         |      ^
      11 |   setTimeout(() => {
      12 |     expect(false).toBe(true);
      13 |     done();

      at new Spec (node_modules/jest-jasmine2/build/jasmine/Spec.js:116:22)
      at Object.<anonymous> (test/myTest.test.ts:10:6)
```

Howeveeeer the 2 recommendations of the doc to either use a `try...catch` block, or return a `Promise` had no effects on this output. Specifically, I tried the following

```ts
test('try...catch #1', done => {
  try {
    setTimeout(() => {
      expect(false).toBe(true);
      done();
    }, 10000);
  } catch (e) {
    done(e);
  }
});

test('try...catch #2', done => {
  setTimeout(() => {
    try {
      expect(false).toBe(true);
      done();
    } catch (e) {
      done(e);
    }
  }, 10000);
});

test('new Promise', () => { // Marking this as `async` made no difference
  return new Promise(done => {
    setTimeout(() => {
      expect(false).toBe(true);
      done();
    }, 10000);
  });
});
```

All those examples resulted in the exact same output.
I'm also wondering the rule for this need, and wondering if Jest didn't use to do this differently in earlier versions, which would have made this rule necessary.

**Comment by G-Rath** [comment]
@SimenB have updated your patch to match the new state of things (`.js` -> `.ts` & the rule now uses `suggestions`)

**Comment by G-Rath** [comment]
@SimenB looking over this, I'm not sure how reasonable it is to support, as there's a whole lot of ways the `done` argument could be used that we'd miss.

**Comment by G-Rath** [comment]
Closing because:
  * this is old
  * this is hard
  * people should stop using `done` 😄

(happy to reopen if we suddenly get an influx of people complaining they're hitting this footgun and wish there was a lint rule)
